docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -0,0 +1,597 @@
+[source,java]
+----
+@BuildStep
+void syntheticObserver(ObserverRegistrationPhaseBuildItem observerRegistrationPhase,
+            BuildProducer<MyBuildItem> myBuildItem, 
+            BuildProducer<ObserverConfiguratorBuildItem> observerConfigurators) {
+   observerConfigurators.produce(new ObserverConfiguratorBuildItem(observerRegistrationPhase.getContext().configure().observedType(String.class)
+                             .notify(mc -> {
+                               // do some gizmo bytecode generation...
+                             }).done();
+   myBuildItem.produce(new MyBuildItem());
+}
+----
+
+[[generated_beans]]
+== Use Case - I Have a Generated Bean Class
+
+No problem.
+You can generate the bytecode of a bean class manually and then all you need to do is to produce a `GeneratedBeanBuildItem` instead of `GeneratedClassBuildItem`.
+
+.`GeneratedBeanBuildItem` Example
+[source,java]
+----
+@BuildStep
+GeneratedBeanBuildItem generatedBean(BuildProducer<GeneratedBeanBuildItem> generatedBeans) {
+    ClassOutput beansClassOutput = new GeneratedBeanGizmoAdaptor(generatedBeans); <1>
+    ClassCreator beanClassCreator = ClassCreator.builder().classOutput(beansClassOutput)
+                .className("org.acme.MyBean")
+                .build();
+    beanClassCreator.addAnnotation(Singleton.class);
+    beanClassCreator.close(); <2>
+}
+----
+<1> A convenient `io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor` can be used to produce a `GeneratedBeanBuildItem` from Gizmo constructs.
+<2> The resulting bean class is something like `public class @Singleton MyBean { }`
+
+== Use Case - I Need to Validate the Deployment
+
+Sometimes extensions need to inspect the beans, observers and injection points, then perform additional validations and fail the build if something is wrong.
+
+=== _Solution 1_: register a `BeanDeploymentValidatorBuildItem`
+
+If you need to validate the deployment with no _build time context_ use the `BeanDeploymentValidatorBuildItem` and leverage the `io.quarkus.arc.processor.BeanDeploymentValidator`.
+
+.`BeanDeploymentValidatorBuildItem` Example
+[source,java]
+----
+@BuildStep
+BeanDeploymentValidatorBuildItem beanDeploymentValidator() {
+    return new BeanDeploymentValidatorBuildItem(new BeanDeploymentValidator() {
+         public void validate(ValidationContext context) {
+             for (InjectionPointInfo injectionPoint : context.getInjectionPoints()) {
+                 // NOTE: You can't produce a build item from within this callback!
+                 System.out.println("Injection point: " + injectionPoint);
+             }
+         }
+    });
+}
+----
+
+TIP: You can easily filter all registered beans via the convenient `BeanStream` returned from the `ValidationContext.beans()` method.
+
+=== _Solution 2_: `ValidationPhaseBuildItem`
+
+If an extension needs to produce other build items during the validation it should use the `ValidationPhaseBuildItem` instead.
```

Comment:
Not for this PR, but this started bugging me in the previous chapters and now I just have to ask: what's the difference between those 2 build items (`BeanDeploymentValidatorBuildItem` and `ValidationPhaseBuildItem`)? It seems to me that the 2nd is stricly more powerful, easier to use and generally feels like a better approach, while the 1st looks like some legacy thing we might wanna deprecate (and possibly even remove) at some point?

Reply:
Well, you're right that it's a result of API evolution ;-). However, the difference is mentioned in _the first sentence_ of the description of both build items... In the beginning there was the `BeanDeploymentValidatorBuildItem` which is basically a callback that is used inside ArC during bootstrap. This works just fine but the problem is that you can't produce a build item from within this callback, ie. outside a `@BuildStep` invocation (well-defined limitation). And so we had to split the bootstrap into multiple phases and introduce the corresponding `Phase` build items. The downside of this approach is that you need to add the corresponding `BuildProducer` otherwise it could be ignored or processed at the wrong time.

Ideally, we should provide something like `SyntheticBeanBuildItem` for all phases. However, I'm not sure it's worth the effort.

Reply:
Yea I read the first sentence, but it didn't quite occur to me that _this particular difference_ warrants 2 mechanisms to exists to achieve the same thing. It just feels strange.

Reply:
> It just feels strange.

I agree. One reason was backwards compatibility. The other reason is that neither of these solutions is "good enough". 

We should probably deprecate the `BeanRegistrarBuildItem` because we have the `SyntheticBeanBuildItem` + `BeanDiscoveryFinishedBuildItem`/`SynthesisFinishedBuildItem` combo.

And we should probably deprecate `BeanDeploymentValidatorBuildItem` and `ObserverRegistrarBuildItem` as well.

Reply:
`BeanRegistrarBuildItem`, `BeanDeploymentValidatorBuildItem` and `ObserverRegistrarBuildItem` are now deprecated. I'm not sure about the `ContextRegistrarBuildItem` - there is actually a chicken-egg problem with scope annotations. I'll try to fix it tomorrow.

Reply:
`ContextRegistrarBuildItem` deprecated too. We had to introduce a new build item to register the scope annotation - `CustomScopeBuildItem`. The API is a bit cumbersome but I have no other idea how to fix the chicken-egg problem with bean defining annotations.